### PR TITLE
Add support for v2Checkpoint in Delta Lake

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Relates to https://github.com/trinodb/trino/issues/19345

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Add support for reading tables with [v2 checkpoint](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#v2-spec). ({issue}`19345`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
Suspending this work to prioritize #19588

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 1]
```diff
@@ -245,7 +250,73 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
...
+                        domainCompactionThreshold,
+                        partitionConstraint,
+                        addStatsMinMaxColumnFilter);
+                result = Optional.of(ImmutableList.copyOf(parquetCheckpoint));
+            }
+            else {
+                throw new IllegalArgumentException("Unsupported v2 checkpoint file format: " + checkpointFile.location());
+            }
+
+            Optional<List<DeltaLakeTransactionLogEntry>> entries = result;
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:288]: Do we need this additional variable? Can `result` be used instead?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 2]
```diff
@@ -245,7 +250,73 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
         catch (IOException e) {
             throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, format("Unexpected IO exception occurred while retrieving the length of the file: %s for the table %s", checkpoint, table), e);
         }
-        return new CheckpointEntryIterator(
+        if (checkpoint.getV2Checkpoint().isPresent()) {
+            Optional<List<DeltaLakeTransactionLogEntry>> result;
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:254]:
```suggestion
            Optional<List<DeltaLakeTransactionLogEntry>> checkpointEntries;
```

Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:254]: Can you encapsulate v2 logic in helper function so it's clearer what happens in case of v1?

**Comment by ebyhr** [comment]
Rebased on master to resolve logical conflicts.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 3]
```diff
@@ -1305,6 +1305,49 @@ private void assertPartitionValuesParsedCondition(String tableName, int id, @Lan
...
+        String tableName = "test_v2_checkpoint_" + randomNameSuffix();
+        Path tableLocation = Files.createTempFile(tableName, null);
+        copyDirectoryContents(new File(Resources.getResource(resourceName).toURI()).toPath(), tableLocation);
+
+        assertUpdate("CALL system.register_table('%s', '%s', '%s')".formatted(getSession().getSchema().orElseThrow(), tableName, tableLocation.toUri()));
+        assertThat(query("DESCRIBE " + tableName))
+                .projected("Column", "Type")
+                .skippingTypesCheck()
+                .matches("VALUES ('a', 'integer'), ('b', 'integer')");
+        assertQuery("SELECT * FROM " + tableName, "VALUES (1, 2)");
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1330]: Any way to test that checkpoint is actually read? Maybe deleting other transaction log files from resources?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import static java.util.Objects.requireNonNull;
+
+public record V2Checkpoint(String path)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java:18]: I don't think there is.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1333]: I think verifying the existence of _last_checkpoint is sufficient. Added another assertion.

**Review by pajaks — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 5]
```diff
@@ -200,7 +203,7 @@ public Stream<DeltaLakeTransactionLogEntry> getCheckpointTransactionLogEntries(
         return getCheckpointPartPaths(checkpoint).stream()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:203]: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#table-features

> A table must not use [multi-part checkpoints](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#multi-part-checkpoint)

Where are we checking this constraint?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java [hunk 6]
```diff
@@ -109,6 +110,13 @@ public CheckpointSchemaManager(TypeManager typeManager)
                 RowType.field("readVersion", BIGINT),
                 RowType.field("isolationLevel", VARCHAR),
                 RowType.field("isBlindAppend", BOOLEAN)));
+
+        sidecarEntryType = RowType.from(ImmutableList.<RowType.Field>builder()
+                .add(RowType.field("path", VARCHAR))
+                .add(RowType.field("sizeInBytes", BIGINT))
+                .add(RowType.field("modificationTime", BIGINT))
+                .add(RowType.field("tags", stringMap))
+                .build());
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java:119]:
```suggestion
        sidecarEntryType = RowType.from(ImmutableList.of(
                RowType.field("path", VARCHAR),
                RowType.field("sizeInBytes", BIGINT),
                RowType.field("modificationTime", BIGINT),
                RowType.field("tags", stringMap)));

```

Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java:119]: in sync with the other initializations from the class.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TransactionLogTail.java [hunk 7]
```diff
@@ -120,6 +120,12 @@ public static Optional<List<DeltaLakeTransactionLogEntry>> getEntriesFromJson(lo
     {
         Location transactionLogFilePath = getTransactionLogJsonEntryPath(transactionLogDir, entryNumber);
         TrinoInputFile inputFile = fileSystem.newInputFile(transactionLogFilePath);
+        return getEntriesFromJson(entryNumber, inputFile);
+    }
+
+    public static Optional<List<DeltaLakeTransactionLogEntry>> getEntriesFromJson(long entryNumber, TrinoInputFile inputFile)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TransactionLogTail.java:126]: optional: This change can be potentially extracted to a preparatory commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 8]
```diff
@@ -259,7 +280,87 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
...
+                            .build(),
+                    metadataEntry,
+                    protocolEntry,
+                    stats,
+                    parquetReaderOptions,
+                    checkpointRowStatisticsWritingEnabled,
+                    domainCompactionThreshold,
+                    partitionConstraint,
+                    addStatsMinMaxColumnFilter);
+            checkpointLogEntries = Optional.of(ImmutableList.copyOf(parquetCheckpoint));
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:329]: Pls attempt materializing the checkpoint entries late and not upfront.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 9]
```diff
@@ -259,7 +280,87 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
...
+        }
+        else {
+            throw new IllegalArgumentException("Unsupported v2 checkpoint file format: " + checkpointFile.location());
+        }
+
+        if (checkpointLogEntries.isEmpty()) {
+            return Stream.empty();
+        }
+        Stream<DeltaLakeTransactionLogEntry> transactionLogs = Stream.empty();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:338]: This algorithm seems to be solvable with streams & flatmap (in a similar fashion to what is being done in https://github.com/trinodb/trino/blob/b90fe09f4f28ee1f2cc6a74283b311972242d673/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java#L200-L217 )

File: plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+Data generated using Delta Lake 3.0.0:
+
+```sql
+SET spark.databricks.delta.checkpointV2.topLevelFileFormat = parquet;
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md:4]:
```suggestion
SET spark.checkpointV2.topLevelFileFormat = parquet;
```

This seems incorrect

https://github.com/delta-io/delta/blob/7d41fb7bbf63af33ad228007dd6ba3800b4efe81/spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala#L743-L754

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java (hunk 6 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java:119]: Using builder is better because it reduces future diff when adding a new entries to the bottom.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+// https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-metadata
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java:21]: That's true, but we want to check the latest protocol. I will skip this comment for consistency with existing protocol comments.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+// https://github.com/delta-io/delta/blob/master/PROTOCOL.md#sidecar-file-information
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java:21]: That's true, but we want to check the latest protocol. I will skip this comment for consistency with existing protocol comments.

File: plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md:4]: Good catch, thanks!

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public V2Checkpoint
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java:20]: Is it intentional that you don't add the other properties? (taken from one of the resource files of this PR)

```
{
   "path":"00000000000000000001.checkpoint.b6785ba8-b035-4760-b814-1f115ccb6167.parquet",
   "sizeInBytes":23272,
   "modificationTime":1708995923000,
   "nonFileActions":[
      {
         "protocol":{
            "minReaderVersion":3,
            "minWriterVersion":7,
            "readerFeatures":[
               "v2Checkpoint"
            ],
            "writerFeatures":[
               "v2Checkpoint"
            ]
         }
      }
   ]
}
```

File: plugin/trino-delta-lake/src/test/resources/databricks133/v2_checkpoint_json/_delta_log/00000000000000000001.checkpoint.ae4ea00c-afe0-4e8b-ad87-28125262bd44.json [hunk 14]
```diff
@@ -0,0 +1,4 @@
...
+{"sidecar":{"path":"00000000000000000001.checkpoint.0000000001.0000000001.59e66709-d0ca-402d-9375-0e3c4fcf8d73.parquet","sizeInBytes":12007,"modificationTime":1708995845000}}
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/resources/databricks133/v2_checkpoint_json/_delta_log/00000000000000000001.checkpoint.ae4ea00c-afe0-4e8b-ad87-28125262bd44.json:2]: it would be nice to have a resource to test reading from a checkpoint with multiple sidecars.
i haven't found any knobs to do this - just found in delta.io code unit tests for writing such checkpoints.
in anycase, optional at this point.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DomainMetadataEntry.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import static java.util.Objects.requireNonNull;
+
+public record DomainMetadataEntry(String domain, String configuration, boolean removed)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DomainMetadataEntry.java:18]: The field already exists. I will change to Boolean from boolean though.

[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DomainMetadataEntry.java:18]: This is optional change. I will remove this class.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java (hunk 13 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/V2Checkpoint.java:20]: I don't remember to be honest. But I think we don't need other fields for now.

File: plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/resources/deltalake/v2_checkpoint_parquet/README.md]: I confirmed `spark.checkpointV2.topLevelFileFormat` doesn't work. The previous `spark.databricks.delta.checkpointV2.topLevelFileFormat` is the right session name. 

https://github.com/delta-io/delta/blob/7d41fb7bbf63af33ad228007dd6ba3800b4efe81/spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala#L33-L37

**Comment by findinpath** [comment]
In case if https://github.com/trinodb/trino/pull/20851 lands first, please add as well a test in `TestDeltaLakeAlluxioCacheFileOperations`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckpointsCompatibility.java [hunk 16]
```diff
@@ -679,6 +679,43 @@ private void testWriteStatsAsStructEnabled(Consumer<String> sqlExecutor, String
...
+        testV2CheckpointMultipleSidecars("parquet");
+    }
+
+    private void testV2CheckpointMultipleSidecars(String format)
+    {
+        String tableName = "test_dl_v2_checkpoint_multiple_sidecars_" + randomNameSuffix();
+        String tableDirectory = "delta-compatibility-test-" + tableName;
+
+        onDelta().executeQuery("SET spark.databricks.delta.checkpointV2.topLevelFileFormat = " + format);
+        onDelta().executeQuery("SET spark.databricks.delta.checkpoint.partSize = 1");
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckpointsCompatibility.java:695]: great - i missed it while skimming delta oss code.

**Review by findinpath — APPROVED** [review]

**Comment by findinpath** [comment]
> In case if #20851 lands first, please add as well a test in `TestDeltaLakeAlluxioCacheFileOperations`

@ebyhr FYI https://github.com/trinodb/trino/pull/20851 landed

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+public record CheckpointMetadataEntry(long version, Optional<Map<String, String>> tags)
+{
+    public CheckpointMetadataEntry
+    {
+        requireNonNull(tags, "tags is null");
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java:26]: should we create immutable copy of the enclosed map here?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+public record SidecarEntry(String path, long sizeInBytes, long modificationTime, Optional<Map<String, String>> tags)
+{
+    public SidecarEntry
+    {
+        requireNonNull(path, "path is null");
+        requireNonNull(tags, "tags is null");
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java:27]: same question

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 19]
```diff
@@ -199,30 +203,27 @@ public Stream<DeltaLakeTransactionLogEntry> getCheckpointTransactionLogEntries(
...
+                        addStatsMinMaxColumnFilter,
+                        fileSystem));
     }
 
     public Optional<Long> getLastCheckpointVersion()
     {
         return lastCheckpoint.map(LastCheckpoint::version);
     }
 
-    private CheckpointEntryIterator getCheckpointTransactionLogEntries(
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:225]: nit: this API change can be a separate commit (no change requested)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 20]
```diff
@@ -233,7 +234,8 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
             LastCheckpoint checkpoint,
             TrinoInputFile checkpointFile,
             TupleDomain<DeltaLakeColumnHandle> partitionConstraint,
-            Optional<Predicate<String>> addStatsMinMaxColumnFilter)
+            Optional<Predicate<String>> addStatsMinMaxColumnFilter,
+            TrinoFileSystem fileSystem)
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:238]: Add fileSystem after session

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 21]
```diff
@@ -260,6 +278,97 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
                 domainCompactionThreshold,
                 partitionConstraint,
                 addStatsMinMaxColumnFilter);
+        return stream(checkpointEntryIterator)
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:281]: wrapping doesn't improve readability. let's move onClose to the same line as stream()

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 22]
```diff
@@ -260,6 +278,97 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
...
+            LastCheckpoint checkpoint,
+            TrinoInputFile checkpointFile,
+            TupleDomain<DeltaLakeColumnHandle> partitionConstraint,
+            Optional<Predicate<String>> addStatsMinMaxColumnFilter,
+            TrinoFileSystem fileSystem,
+            long fileSize)
+    {
+        return getV2CheckpointEntries(session, entryTypes, metadataEntry, protocolEntry, checkpointSchemaManager, typeManager, stats, checkpoint, checkpointFile, partitionConstraint, addStatsMinMaxColumnFilter, fileSystem, fileSize)
+                .flatMap(entry -> {
+                    if (entry.getSidecar() == null) {
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:302]: there is a better way of doing this called .mapMulti that is used like that:

```
.mapMulti((entry, builder) -> {
                    if (entry.getSidecar() == null) {
                        builder.accept(entry);
                        return;
                    }
                    Location sidecar = checkpointFile.location().sibling("_sidecars").appendPath(entry.getSidecar().path());
                    CheckpointEntryIterator iterator = new CheckpointEntryIterator(
                            fileSystem.newInputFile(sidecar),
                            session,
                            fileSize,
                            checkpointSchemaManager,
                            typeManager,
                            entryTypes,
                            metadataEntry,
                            protocolEntry,
                            stats,
                            parquetReaderOptions,
                            checkpointRowStatisticsWritingEnabled,
                            domainCompactionThreshold,
                            partitionConstraint,
                            addStatsMinMaxColumnFilter);

                    for (CheckpointEntryIterator it = iterator; it.hasNext();) {
                        builder.accept(it.next());
                    }
                });
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 23]
```diff
@@ -538,6 +546,31 @@ private DeltaLakeTransactionLogEntry buildRemoveEntry(ConnectorSession session,
         return DeltaLakeTransactionLogEntry.removeFileEntry(result);
     }
 
+    private DeltaLakeTransactionLogEntry buildSidecarEntry(ConnectorSession session, int pagePosition, Block block)
+    {
+        log.debug("Building sidecar entry from %s pagePosition %d", block, pagePosition);
+        if (block.isNull(pagePosition)) {
+            return null;
+        }
+        int sidecarFields = 4;
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:555]: I don't understand where 4 comes from

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java [hunk 24]
```diff
@@ -111,6 +112,13 @@ public CheckpointSchemaManager(TypeManager typeManager)
                 RowType.field("partitionValues", stringMap),
                 RowType.field("deletionTimestamp", BIGINT),
                 RowType.field("dataChange", BOOLEAN)));
+
+        sidecarEntryType = RowType.from(ImmutableList.<RowType.Field>builder()
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java:116]: could this be a const?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java [hunk 25]
```diff
@@ -153,9 +167,23 @@ public CdcEntry getCDC()
...
+    public CheckpointMetadataEntry getCheckpointMetadata()
+    {
+        return checkpointMetadata;
+    }
+
     public DeltaLakeTransactionLogEntry withCommitInfo(CommitInfoEntry commitInfo)
     {
-        return new DeltaLakeTransactionLogEntry(txn, add, remove, metaData, protocol, commitInfo, cdcEntry);
+        return new DeltaLakeTransactionLogEntry(txn, add, remove, metaData, protocol, commitInfo, cdcEntry, sidecar, checkpointMetadata);
     }
```
Comment (Praveen2112) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java:188]: A basic question - Don't we need the sidecard and checkpointMetadata for hashcode and equals check ? If not can we add a reason as a comment ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+// https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-metadata
+public record CheckpointMetadataEntry(long version, Optional<Map<String, String>> tags)
+{
+    public CheckpointMetadataEntry
```
Comment (Praveen2112) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/CheckpointMetadataEntry.java:24]: Do we need any additional validation of the version like non-zero ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+// https://github.com/delta-io/delta/blob/master/PROTOCOL.md#sidecar-file-information
+public record SidecarEntry(String path, long sizeInBytes, long modificationTime, Optional<Map<String, String>> tags)
```
Comment (Praveen2112) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/SidecarEntry.java:22]: Can we have any validation on non-zero for sizeInBytes ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 23 above)
Comment (Praveen2112) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:555]: I think side car has 4 entries 
```
{
  "sidecar":{
    "path": "016ae953-37a9-438e-8683-9a9a4a79a395.parquet",
    "sizeInBytes": 2304522,
    "modificationTime": 1512909768000,
    "tags": {}
  }
}
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAlluxioCacheFileOperations.java
Comment (Praveen2112) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAlluxioCacheFileOperations.java:1]: Shouldn't we add a similar test for `TestDeltaLakeFileOperations` ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 23 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:555]: Right, 4 comes from field numbers of `sidecar`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java (hunk 24 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointSchemaManager.java:116]: It uses `stringMap` that depends on injected TypeManager.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java (hunk 19 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:225]: Let me skip this comment as separated commit doesn't improve diff so much.

**Comment by ebyhr** [comment]
Addressed comments.

**Review by pajaks — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 28]
```diff
@@ -538,6 +546,31 @@ private DeltaLakeTransactionLogEntry buildRemoveEntry(ConnectorSession session,
...
+        SqlRow sidecarEntryRow = getRow(block, pagePosition);
+        log.debug("Block %s has %s fields", block, sidecarEntryRow.getFieldCount());
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:557]: do we need this log?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 29]
```diff
@@ -538,6 +546,31 @@ private DeltaLakeTransactionLogEntry buildRemoveEntry(ConnectorSession session,
...
+                    format("Expected block %s to have %d children, but found %s", block, sidecarFields, sidecarEntryRow.getFieldCount()));
+        }
+        RowType type = sidecarType.orElseThrow();
+        CheckpointFieldReader sidecar = new CheckpointFieldReader(session, sidecarEntryRow, type);
+        SidecarEntry result = new SidecarEntry(
+                sidecar.getString("path"),
+                sidecar.getLong("sizeInBytes"),
+                sidecar.getLong("modificationTime"),
+                Optional.ofNullable(sidecar.getMap(stringMap, "tags")));
+        log.debug("Result: %s", result);
```
Comment (wendigo) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:570]: do we need this log?

**Review by wendigo — APPROVED** [review]

**Comment by wendigo** [comment]
@ebyhr merge at will :)
